### ISSUE TITLE
AD smoke test for compute_msl_s_matrix passes on a ZERO gradient — it does not gate differentiability

### ISSUE THREAD

**Issue by bk-squared** [bk-squared/rfx#515]
Found while checking whether #511/#507's extractor change affects the AD path. **Pre-existing on main**, not a regression from that work.

## The test cannot fail for the reason it exists

`tests/test_msl_sparam_ad.py::test_compute_msl_s_matrix_ad_smoke_has_finite_gradient` asserts only:

```python
grad = jax.grad(objective)(jnp.asarray(0.7, dtype=jnp.float32))
assert bool(jnp.isfinite(grad))
assert not bool(jnp.isnan(grad))
```

Measured on main **and** on `fix/511-msl-modal-voltage-off-by-one`:

```
[AD smoke] grad = 0.000000e+00
1 passed
```

`0.0` is finite and not NaN, so a **completely severed tape passes**. The docstring says the test "gates only that the assembly is differentiable" — it does not do that.

## Where theta actually enters, and what responds

`_fake_run_for_theta` (`:255-285`) injects `theta` at exactly one place: `run_idx == 0 and port_idx == 1`, i.e. the receive port during drive 0. That is the `b_2`/`a_2` path, so `S[1,0]` should depend on it under both the old single-ratio rule and the new multi-drive solve.

Measured sensitivities at `theta = 0.7` (branch; the test's own objective is the first row):

| objective | d/dtheta |
|---|---|
| `Re S21` — **what the test uses** | `+0.000000e+00` |
| `Im S21` | `+0.000000e+00` |
| `abs S21` | `+0.000000e+00` |
| `Re S11` | `+1.299351e-08` |
| `abs S11` | `+1.299351e-08` |
| `sum abs S` | `+1.333008e-08` |

So the tape is not entirely dead — S11 responds — but at `~1.3e-8` against `S ~ O(1)` and `theta = 0.7`, that is f32 rounding, not a derivative. And the component the test picked is exactly zero.

I did not localise where theta's contribution dies. Ruled out: passivity projection is skipped under tracing (`_sparams.py`, `not is_tracer(S)` guard), so it is not the SVD.

## Why it matters now

The real AD verification for this lane, `tests/test_msl_ad_fd_converged.py`, is marked **both** `@pytest.mark.gpu` and `@pytest.mark.slow` — so it runs in neither the fast lane nor the CPU slow lane, only in the VESSL GPU suite. With the smoke test unable to fail, **there is no CPU-runnable AD coverage of `compute_msl_s_matrix` at all**.

That is what makes #511/#507's effect on gradients unverifiable locally: I can say the change did not obviously break AD, and I cannot say it preserved gradient accuracy. PR #486's "0.011% AD-vs-FD converged" claim needs the GPU suite re-run before it can be repeated post-change.

## Suggested fix

1. Pick an objective with a **non-zero** analytic sensitivity to `theta`, and assert a magnitude, not just finiteness — e.g. `assert abs(grad) > 1e-4` with the bound derived from the synthetic construction rather than from a measured run.
2. Add the mutation twin: `jax.lax.stop_gradient` the wave amplitudes and assert the test goes red. A differentiability test that passes with the tape cut is the defect being fixed here, so the fix needs to prove it does not.
3. Consider a CPU-runnable AD-vs-FD check on a small real fixture, so the lane is not GPU-only for its only real gradient gate. `test_msl_ad_fd_converged.py` is `gpu` because of runtime, not because gradients need a GPU.
4. Separately worth understanding: why `Re S21` has an exactly-zero sensitivity to a parameter that scales the receive port's `alpha`. That may be a synthetic-fixture artifact, or it may be real information about the assembly.

**Comment by bk-squared** [comment]
## Correction: the zero gradient is PR-introduced, not inherited — this issue's premise was wrong

The independent review of PR #516 caught a false claim in this issue and in the PR body,
and I have now reproduced the correction directly:

- **clean main** (`75d84fa`): `[AD smoke] grad = 3.912601e-01`
- **PR #516 branch**: `[AD smoke] grad = 0.000000e+00`

The test file is byte-identical between the two, so the change is entirely the extractor.
My original claim here — *"Measured on main and on the branch: grad = 0.0 ... so it is not a
regression"* — came from a broken comparison: both invocations ran inside the worktree
because the `cd` at the start of a single shell call persisted across the `;`-separated
second pytest. I compared the branch against itself.

**What stays true, and matters just as much:** the review also measured the forward
objective at θ = 0.6 / 0.7 / 0.8 = −2.85e-08 / 0.0 / −2.85e-08 — so the finite-difference
slope is also ~0 under the new assembly. The multi-drive solve maps this synthetic fixture
to `Re(S21) ≈ 0` with ≈0 sensitivity, and the test asserts only `isfinite`/`not isnan`. So
the test's **vacuity** (it passes on a dead value) is real either way — but it was
*introduced by PR #516's assembly change*, not inherited from main, and the fix must both
rebuild the fixture for genuine θ-sensitivity AND assert a nonzero gradient floor.

Unit-level AD is not implicated: the review measured `jax.grad` through
`msl_solve_s_from_waves` against central FD at 1.165e-9 relative (f64), and a skeptic's
independent fixture at 1.7e-10. The gap is the smoke *fixture*, not the solve.

Also for the record: this makes the smoke test's history worth re-reading — on main the
0.391 value was a live signal, and PR #516 silently kills it. The end-to-end
`test_msl_ad_fd_converged` (gpu+slow, in no scheduled lane) is now the only place a real
MSL gradient regression could surface.

**Comment by bk-squared** [comment]
Body-level correction pointer: this issue's opening line still reads **"Pre-existing on main, not a regression"** in bold. That is refuted — clean main prints `grad = 3.912601e-01`, the PR branch printed `0.000000e+00`; the vacuity was introduced by the assembly change, not inherited. The correction is in the 2026-07-30 comment below; the body has been left as originally written so the record shows what was claimed.

Post-merge addition: the support matrix's **"converged f64 referee 0.000110"** AD-validation claim (`sparameter_support_matrix.md:44` and `:137-141`, plus the `.json` `evidence` bullet and `ad_evidence` field) is a pre-`f95240f` measurement. Gradients now flow through the multi-drive solve and no post-merge VESSL GPU run exists (newest log 2026-07-28). Re-measurement of that number lands here, after the next `scripts/vessl_gpu_suite.yaml` run; the caveat annotation is tracked in #520.

**Comment by bk-squared** [comment]
Re-measurement done: **it is worse, and it is now a hard failure.** VESSL 369367250602 (`scripts/vessl_gpu_suite.yaml`, main@`f95240f`, 2026-07-31, 1:07:41):

```
1 failed, 193 passed, 70 skipped, 2 xfailed
FAILED tests/test_msl_ad_fd_converged.py::test_msl_ad_fd_converged_tight
rel_err = 0.8519  (threshold 0.10)   g_ad = -4.2362e-03   g_fd = -2.8610e-02
```

One failure in the entire GPU suite, and it is the gradient gate. Pre-#516 the same gate read `rel_err = 0.00011`. Split out as **#527** with the full instrumentation, the candidate mechanisms, and the cheapest first probe (an h-sweep, to bound how much of the 0.85 is FD reference noise before chasing the AD path).

That settles this issue's open half: the "converged f64 referee 0.000110" number is not merely stale, it is **contradicted** on current main. The support-matrix annotation tracked in #520 should now say so rather than 'pending re-measurement'.

The smoke-test half of this issue (the test that passes on grad = 0.0) is unchanged and still wants the fixture rebuild + nonzero-gradient floor.

**Comment by bk-squared** [comment]
**Status update with fresh measurements (2026-08-03):**

1. **The VESSL-rerun leg is DONE.** The rebuilt f64 AD-vs-FD gate (post-#516/#529 comparator) ran on its owner platform today (run 369367251600, `scripts/vessl_msl_ad_gate.yaml`): **rel_err = 0.0331 vs the 0.10 threshold — PASS, 3 passed in 140 s** (gate + resolving-power falsifier + dtype lock). So differentiability of the real objective is verified post-extractor-change on GPU. (Yesterday's full gpu-suite run 369367251342 was also green: 194 passed / 0 failed.)

2. **The smoke-test defect is still live, and now precisely localized.** Measured on today's main (fd2325e lineage): `test_compute_msl_s_matrix_ad_smoke_has_finite_gradient` prints `grad = 0.000000e+00` and passes — the assertion set (isfinite, not-NaN) cannot fail on a severed tape. Since the tight gate proves the AD path itself is healthy, the zero means the SMOKE'S OBJECTIVE is structurally flat on main — which is #530's finding (the objective differentiates a passivity-pinned quantity) showing up at smoke scale.

**Narrowed remaining scope**: rebuild the smoke around an objective with a measured non-zero gradient (per #530: choose with measured ULP headroom — the tight gate's objective at reduced fixture scale is the natural candidate) and assert |grad| > a measured floor, not merely finiteness. This issue and #530 should likely be resolved by the same objective decision.
